book/src/solana-scripts.md

```diff
@@ -0,0 +1,253 @@
+following transfers.  The script execution succeeds if and only if the script
+generates the exact same instruction vector during execution.  Bob and Alice
+have no way to ensure what the state of any of the programs will be during the
+start of the script. The explicit instruction vector ensures that the script
+behaves as an infallible TTP.
+
+* `SystemInstruction::Create{...}` - This instruction references the keys
+generated by the script, and therefore cannot be signed by the user.  During the
+script execution the script will add the signature to the instruction.
+
+## Script Example
+
+In this example, a script accepts tokens from two different accounts, and pays
+out the total to whoever wins the game of Tic-Tac-Toe.
+
+```
+enum BetOnTicTacToeScript {
+    Initialize {amount: u64, game: Pubkey},
+    Claim,
+};
+
+// This is an additional program that implements some features that this script
+// needs.  Scripts cannot mutate state directly, and rely on programs to handle
+// state mutation.
+const BetOnTicTacToeProgramId: Pubkey = [];
+enum BetOnTicTacToeProgramInstruction {
+    // copy the game id from to account 1
+    StoreGame,
+}
+
+pub fn process_instruction(
+    program_id: &Pubkey,
+    keyed_accounts: &mut [KeyedAccount],
+    data: &[u8],
+) -> Result<(), InstructionError> {
+    let cmd = deserialize(&data)?;
+    match cmd {
+        case BetOnTicTacToeScriptInstruction::Initialize{ amount, game} => {
+            //The scripts system account to store lamports
+            let script_tokens_key = script::keypair_pubkey(0);
+
+            let from_alice = system_instruction::transfer(
+                            keyed_accounts[1].key, //alice
+                            script_tokens_key,
+                            amount);
+
+            //alice must have signed this instruction
+            script::process_instruction(from_alice)?;
+            let from_bob = system_instruction::transfer(
+                            keyed_accounts[2].key, //bob
+                            script_tokens_key,
+                            amount);
+
+            //bob must have signed this instruction
+            script::process_instruction(from_bob)?;
+
+            //The scripts `program_id` account to store the game state
+            let script_data_key = script::keypair_pubkey(1);
+
+            //to save the game, the data key needs to be allocated
+            let mut create = system_instruction::create(
+                            script_token_key,
+                            script_data_key,
+                            2,
+                            size_of(game),
+                            BetOnTicTacToeProgramId);
+            script::sign_instruction(&mut create, 0);
+            script::process_instruction(create)?;
+
+            let prize = amount * 2 - 2;
+            //call BetOnTicTacToeProgramId
+            let store_game = betontictactoe::store_game(
```

Comment:
Maybe I'm missing something.  Serialization of the game info requires that the above `system_create` has succeeded (an account was created to store the data).  Are all the instructions issued by the script executed first in some sandbox and then committed for real at script completion?  If not that implies that each instruction is executed fully during script execution and if there are any failures those executed instructions would have to be unwound...

Reply:
@jackcmay when script calls `process_instruction`, it yields to the runtime, and the runtime checks that the issued instruction matches the instruction in the vector, then calls it, and resumes the script once the issued instruction is done.

Reply:
If an instruction in the scripts fails, what to do about the already "done" instructions?

Reply:
@jackcmay the whole transaction fails.  so those updates are not committed back to the state.